osu.Game.Mode.Osu/Objects/Slider.cs

```diff
@@ -2,18 +2,27 @@
 //Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
 
 using System.Collections.Generic;
+using osu.Game.Database;
 using OpenTK;
+using osu.Game.Beatmaps;
+using System;
 
 namespace osu.Game.Modes.Osu.Objects
 {
     public class Slider : OsuHitObject
     {
-        public override double EndTime => StartTime + (RepeatCount + 1) * Curve.Length;
+        public override double EndTime => StartTime + RepeatCount * Curve.Length / Velocity;
+
+        public double Velocity;
+
+        public override void SetDefaultsFromBeatmap(Beatmap beatmap)
+        {
+            Velocity = 100 / beatmap.BeatLengthAt(StartTime, true) * beatmap.BeatmapInfo.BaseDifficulty.SliderMultiplier;
+        }

```

Comment:
what about when a hitobject needs to get bpm at multiple points in time? what about mod status? what about user settings? there's going to be a whole heap of data which are needed for hitobjects in general. should we centralise all this into a class which gets passed in?